src/protocol/graphics.ts

```diff
@@ -423,3 +416,43 @@ export async function getFirstMeaningfulPaint(limit: number = 10) {
     }
   }
 }
+
+// precache at least this many milliseconds
+const precacheMinTime = 5000;
+// precache at least this many screenshots
+const precacheMinCount = 20;
+// startIndex of the currently running precacheScreenshots() call
+let precacheStartIndex = -1;
+
+export function precacheScreenshots(startTime: number): UIThunkAction {
+  return async ({ dispatch, getState }) => {
+    await paintPointsWaiter;
+
+    const startIndex = mostRecentIndex(gPaintPoints, startTime);
+    if (startIndex === undefined || precacheStartIndex === startIndex) {
+      return;
+    }
+    precacheStartIndex = startIndex;
+
+    let endIndex = mostRecentIndex(gPaintPoints, startTime + precacheMinTime) || startIndex;
+    endIndex = Math.max(endIndex, startIndex + 1 + precacheMinCount);
+    endIndex = Math.min(endIndex, gPaintPoints.length);
+    for (let i = startIndex + 1; i < endIndex; i++) {
```

Comment:
The single-paint increment here is going to issue a lot of fetches during precaching.  I want to suggest computing the indexes of the loop by doing a fixed time-offset (e.g. 10ms increments), and recomputing the gPaintPoints index for each of those times, and then caching the images at those times only.

The only problem with that is synchronizing this with the playback engine so that later precache attempts for overlapping time segments also query the same points.

Can we have a global PLAYBACK_SNAP variable for now, set to 50ms?  Then, within `getGraphicsAtTime`, we snap the time as follows:

```
if (forPlayback) {
  time += PLAYBACK_SNAP - (time % PLAYBACK_SNAP);
}
...
```

This will ensure that for playback, every recording of some length will always request the same paint indexes.